### PR TITLE
Pre SIP: Improvements to Modularity

### PR DESCRIPTION
**Pull request by odersky**
For a tracked class parameter we add a refinement in the constructor type that
the class member is the same as the parameter. E.g.
```scala
class C { type T }
class D(tracked val x: C) { type T = x.T }
```
This will generate the constructor type:
```scala
(x1: C): D { val x: x1.type }
```
Without `tracked` the refinement would not be added. This can solve several problems with dependent class
types where previously we lost track of type dependencies.

In particular it addresses the parser combinator test cases that were prompted by this discussion in Scala Users: https://users.scala-lang.org/t/create-an-instance-of-a-type-class-with-methods-depending-on-type-members/9613/4.

This PR is essentially a feasibility study. It shows that we can get better tracking of types through constructors without breaking anything. 

But I am not sure yet a new modifier `tracked`  is the best way to achieve this. Ideally, we would detect when a class uses a class parameter as a path in a type, and refine those parameters only. The problem is that we don't know this in general when we compute the type of a constructor application, since types of class members are computed lazily on first access. 

Example:
```scala
    trait A:
       type T
       def m: T

    class D(x: A):
       def f = x.m
```
We need to typecheck the RHS of `f` in `D` to figure out that `f`'s result type `x.T` refers to the parameter `x`, so that `x` needs a refinement. So computing precise info of what needs to be refined is infeasible.

Some of the other options are:

 1. The current scheme: use a modifier such as `tracked` to indicate refinement. To make this more robust we could make it an error if a type in a public signature refers to an untracked class parameter. So in the example above, we could enforce that class `D` is declared as `class D(tracked x: A)`.
 2. Overapproximate wildly: Make all class parameters tracked (probably this would cause lots of problems).
 3. Overapproximate more conservatively: Maybe make all non-private class parameters tracked. This is appealing at first glance since private class parameters cannot appear in refinements anyway. On the other hand, it would still make all case class parameters tracked, which is probably overkill again. Variations: Make all parameters declared with explicit `val`s tracked, and/or make all parameters in `given ... with` clauses tracked. 
 4. Try to guess: Assume `tracked` for parameters that appear "obviously" in a path. E.g. there is a path with the name of the parameter in an explicitly given signature in the class. Provide a fallback such as a `tracked` modifier or an annotation for parameters that are not obviously in a path. The goal of this compared to (1) is to be less annoying. Why insist on explicit `tracked` if it is clear from the source that a parameter needs to be refined? But I don't like the fragility of this scheme.

### CODE REVIEW

**Comment by odersky** [comment]
These are variations of a [Scala Users discussion](https://users.scala-lang.org/t/create-an-instance-of-a-type-class-with-methods-depending-on-type-members/9613/4. )

The critical part is shown in parsercombinators-expanded.scala. definition of `combine`:
```scala
def combine[A, B](
      _f: Combinator[A],
      _s: Combinator[B] { type Context = _f.Context}
    ): combine[A, B] {
      type Context = _f.Context
      type Element = (_f.Element, _s.Element)
    } = new combine[A, B](_f, _s).asInstanceOf
    // cast is needed since the type of new combine[A, B](_f, _s)
    // drops the required refinement.
```
We'd need to appy the scheme for instantiating dependent methods also to class instance creations. 
If a new expression of a class that defines type members mentioning parameters, we need to add a refinement
that substitutes actual arguments for formal parameter names.

**Comment by odersky** [comment]
See also https://github.com/lampepfl/dotty/issues/3964

**Comment by odersky** [comment]
I  now see that #3936 is _very_ similar to this PR (and might  also solve some problems wrt inheritance in this PR). Definitely worth following up. Maybe we should re-open #3936? /cc @liufengyun

**Comment by liufengyun** [comment]
> I now see that #3936 is _very_ similar to this PR (and might also solve some problems wrt inheritance in this PR). Definitely worth following up. Maybe we should re-open #3936? /cc @liufengyun @julienrf

Indeed it's similar. Glad to see there's real requirement from real-world programming. This PR is much more polished than #3936, it makes sense to keep the current PR. 

BTW, I remember @mbovel is working on something related.

**Comment by odersky** [comment]
I was particularly interested in the way #3936 handles base types. This PR has nothing in that department. Maybe that's needed to fix the remaining failures in neg/i3964.scala?

**Comment by liufengyun** [comment]
> I was particularly interested in the way #3936 handles base types. This PR has nothing in that department. Maybe that's needed to fix the remaining failures in neg/i3964.scala?

I forgot exactly why the changes in handling base types, but it seems to be motivated by the last test case there. However, I'm not sure if the changes make sense, as I was not very familiar with the base type computation at the time.

**Comment by odersky** [comment]
Some more info: Making all explicit vals tracked causes a failing bootstrap and ~50 failing tests. Particular issues:

 - Private vals escaping in inferred refinements of non-private values
 - Cyclic references between a class and its parents
 - Type errors because a too precise type was inferred

So, not sure this idea will fly.

**Comment by liufengyun** [comment]
I was wondering will the errors go away if `track` is opt-in.

> Some more info: Making all explicit vals tracked causes a failing bootstrap and ~50 failing tests. Particular issues:
> 
> * Private vals escaping in inferred refinements of non-private values

It seems this will go away if `track` is only allowed on non-private values.

> * Cyclic references between a class and its parents
> * Type errors because a too precise type was inferred

This might go way as well with explicit `track` --- at least users can understand and easily fix if too precise inference due to `track` becomes a problem.

> 
> So, not sure this idea will fly.

**Comment by odersky** [comment]
@liufengyun Yes, I agree, explicit `tracked` solves most of these things. But #19015 is still useful since it might us help discover problems that also arise on widespread uses of explicit `tracked` parameters, and that could be fixed now.

**Comment by odersky** [comment]
Another thing that currently does not work is this test from `neg/i3964.scala`:
```scala
object Test3:
  trait Vec(tracked val size: Int)
  class Vec8 extends Vec(8):
    val s: 8 = size    // error, but should work
```
The problem is we currently keep track of refinements in constructors and therefore in constructed values. But we don't do the same thing inside classes. Inside classes we could experiment with

 1. Synthetic abstract members, like in `Vec8` we'd need a synthetic abstract member `val size: 8`. The general case looks more tricky, though.
 2. Refinements on the self type. I.e. `Vec8` would be expanded to `class Vec8 { this: Vec8 { val self: 8 }}`. We do a similar thing for opaque type aliases.

**Comment by liufengyun** [comment]
> But we don't do the same thing inside classes. Inside classes we could experiment with
> 
> 1. Synthetic abstract members, like in `Vec8` we'd need a synthetic abstract member `val size: 8`. The general case looks more tricky, though.
> 2. Refinements on the self type. I.e. `Vec8` would be expanded to `class Vec8 { this: Vec8 { val self: 8 }}`. We do a similar thing for opaque type aliases.

I was thinking whether it is possible to handle it in base type computation logic? Conceptually, it seems to be the same problem as refining types of members with regards to type parameters and prefixes.

**Comment by sjrd** [comment]
I don't think `baseType` is the right solution for this. `baseType` deals with type parameters and prefixes of parent class types because those things are not otherwise accessible. Also they have to deal with variance. For type and value members, we can select them from outside, and we get refined types through `asSeenFrom`, not `baseType`.

Also, `baseType` is currently guaranteed, by spec, to always return a type of the form `p.C[T1, ..., Tn]` (or `p.C` for `n = 0`). This assumption is relied upon in other parts of the spec that call `baseType`. It seems unwise to break that contract without thorough analysis of the implications.

**Comment by liufengyun** [comment]
> > I was thinking whether it is possible to handle it in base type computation logic? Conceptually, it seems to be the same problem as refining types of members with regards to type parameters and prefixes.
> 
> That's an interesting idea, but I am not sure about how this would work in detail?

I was thinking that in the following example the parent call `Vec(8)` gets a refinement type `Vec { val size: 8 }` from the constructor call. Then in base type computation, it takes that refinement into consideration.

```Scala
  trait Vec(tracked val size: Int)
  class Vec8 extends Vec(8):
    val s: 8 = size    // error, but should work
```
My knowledge is limited regarding the base type computation, just an intuition, so not sure if that would work.

**Comment by odersky** [comment]
This PR now bundles three improvements to modularity, all available under the `experimental.modularity` language import:

 - tracked type parameters
 - refinements in class parents
 - non-final type export forwarders

The doc page for these changes is
 https://github.com/lampepfl/dotty/blob/afefb9e02ceeca13ed4c1b6ee0a2fbf603df8e76/docs/_docs/reference/experimental/modularity.md

**Comment by mushtaq** [comment]
How about using `dep` instead of `tracked val`?

```scala
class D(dep x: C) { type T = x.T }
```

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -622,6 +629,10 @@ object RefChecks {
         mbr.isType
         || mbr.isSuperAccessor // not yet synthesized
         || mbr.is(JavaDefined) && hasJavaErasedOverriding(mbr)
+        || mbr.is(Tracked)
+          // Tracked members correspond to existing val parameters, so they don't
+          // count as deferred. The val parameter could not implement the tracked
+          // refinement since it usually has a wider type.
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:635]: Reminds me of the `erased` flag. Might be possible to generalize in the future.

File: docs/_docs/reference/experimental/modularity.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+class C:
+  type T
+  def m(): T
+
+type R = C:
+  type T = Int
+  def m(): 22
+
+class D extends R:
+  def next(): D
```
Comment (liufengyun) [inline docs/_docs/reference/experimental/modularity.md:158]: Should class `C` and `D` be `abstract`?

**Review by liufengyun — APPROVED** [review]
LGTM, nice generalization to parent refinements  🎉

**Comment by liufengyun** [comment]
> How about using `dep` instead of `tracked val`?
> 
> ```scala
> class D(dep x: C) { type T = x.T }
> ```

`dep` is too cryptic as a name --- a full word is better. `dependent` has been discussed [before](https://github.com/lampepfl/dotty/issues/3920#issuecomment-360779916) --- `tracked` is shorter and nicer.

**Comment by LPTK** [comment]
Why not `refined`?

`tracked` evokes the tracking of free variables in capture-tracking. That's the first thing that came to my mind when I saw the keyword.

**Comment by odersky** [comment]
`refined` works on the class, but not on the parameter (and we need it to work on the parameter, in the end). `dependent` is too much type theory jargon, it sounds scary. 

`tracked` in its informal meaning of capture tracking and in its meaning here is actually the same thing. I f a field of a class was a capability, I would expect it to be tracked, in the sense that we keep track of what the actual capability argument was.

**Comment by odersky** [comment]
Needs minor release since it introduces a new Tasty tag.

**Comment by smarter** [comment]
Another possible keyword here would be `precise` since at least in spirit this reminds me of https://github.com/scala/improvement-proposals/pull/48 (/cc @soronpo)

**Comment by soronpo** [comment]
> Another possible keyword here would be `precise` since at least in spirit this reminds me of [scala/improvement-proposals#48](https://github.com/scala/improvement-proposals/pull/48) (/cc @soronpo)

This feature doesn't solve most of the problems the Precise Type SIP targeted so I don't see what is so `precise` about it. I think `dependent` fits better and I know nothing about the theory and I'm not scared :) . 
Other suggestion: `preserved`

**Comment by mushtaq** [comment]
> `dep` is too cryptic as a name --- a full word is better. `dependent` has been discussed [before](https://github.com/lampepfl/dotty/issues/3920#issuecomment-360779916) --- `tracked` is shorter and nicer.

`dep` is [adopted by scala-cli](https://scala-cli.virtuslab.org/docs/reference/directives/#dependency) to mean a `dependency`. In that sense, it is well understood.

My other comment was about the need to write both `tracked` and `val` given that `tracked` is proposed to work only with `val` parameters.

**Comment by odersky** [comment]
This is part of #20061

**Pull request closed without merging**